### PR TITLE
Add VFR detection

### PR DESCRIPTION
**Pull request by awolant**
## Category:

New Feature

## Description:

It adds detection of whether decoded video file is variable frame rate or not and ability to access this information. It is done based on the video file index. we go through the index and check time differences between the frames. If there are any differences decoded file is VFR. 

This check will be optimized in the future. Goal for this PR was to add a check that works in every scenario.

### Affected modules and functionalities:

`FramesDecoder` and its tests.

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2798

### CODE REVIEW

**Comment by awolant** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder_test.cc [hunk 1]
```diff
@@ -36,6 +36,7 @@ class FramesDecoderTestBase : public VideoTestBase {
     ASSERT_EQ(decoder.Width(), ground_truth.Width());
     ASSERT_EQ(decoder.Channels(), ground_truth.NumChannels());
     ASSERT_EQ(decoder.NumFrames(), ground_truth.NumFrames());
+    ASSERT_EQ(decoder.IsVfr(), ground_truth.IsVfr());
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_test.cc:39]: Can you add a test for a case with non-vfr video file?

File: dali/operators/reader/loader/video/frames_decoder.cc [hunk 2]
```diff
@@ -134,6 +135,23 @@ void FramesDecoder::BuildIndex() {
...
 }
 
+void FramesDecoder::DetectVfr() {
+  if (NumFrames() < 3) {
+    is_vfr_ = false;
+    return;
+  }
+
+  int pts_step = index_[1].pts - index_[0].pts;
+  for (int frame_id = 2; frame_id < NumFrames(); ++frame_id) {
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder.cc:145]: How fast is that? Can't we do this when the index is created?

File: dali/operators/reader/loader/video/frames_decoder.cc [hunk 3]
```diff
@@ -134,6 +135,23 @@ void FramesDecoder::BuildIndex() {
...
+    if ((index_[frame_id].pts - index_[frame_id - 1].pts) != pts_step) {
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder.cc:146]: Just to clarify. PTS is an integer and we are not affected by floating point operation precision.

File: dali/operators/reader/loader/video/frames_decoder.cc (hunk 2 above)
Comment (awolant) [reply dali/operators/reader/loader/video/frames_decoder.cc:145]: We could mix it with index creation. For now, I decided to do it like this for readability (look PR teamplate). I added task to optimize it further, when we have proper benchmark. Part of this optimization will be to get this info without any decoding when possible.

File: dali/operators/reader/loader/video/frames_decoder.cc (hunk 3 above)
Comment (awolant) [reply dali/operators/reader/loader/video/frames_decoder.cc:146]: Yes

File: dali/operators/reader/loader/video/frames_decoder_test.cc (hunk 1 above)
Comment (awolant) [reply dali/operators/reader/loader/video/frames_decoder_test.cc:39]: This test covers both VFR and CFR files.

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_test.cc:39]: Ok. Missed that.

File: dali/operators/reader/loader/video/frames_decoder.cc (hunk 2 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder.cc:145]: Understood.

**Review by mzient — APPROVED** [review]

**Pull request merged**